crates/pipeline_manager/src/pipeline_manager.rs

```diff
@@ -603,96 +594,98 @@ struct ProgramCodeResponse {
 #[utoipa::path(
     responses(
-        (status = OK, description = "Program data and code retrieved successfully.", body = ProgramCodeResponse),
-        (status = BAD_REQUEST
-            , description = "Specified program id is not a valid uuid."
+        (status = OK, description = "Programs retrieved successfully.", body = [ProgramDescr]),
+        (status = NOT_FOUND
+            , description = "Specified program name does not exist in the database."
             , body = ErrorResponse
-            , example = json!(example_invalid_uuid_param())),
+            , example = json!(example_unknown_name())),
         (status = NOT_FOUND
             , description = "Specified program id does not exist in the database."
             , body = ErrorResponse
             , example = json!(example_unknown_program())),
     ),
-    params(
-        ("program_id" = Uuid, Path, description = "Unique program identifier")
-    ),
+    params(ProgramIdOrNameQuery, WithCodeQuery),
     tag = "Program"
 )]
-#[get("/program/{program_id}/code")]
-async fn program_code(
+#[get("/programs")]
+async fn get_programs(
     state: WebData<ServerState>,
     tenant_id: ReqData<TenantId>,
-    req: HttpRequest,
+    req: web::Query<ProgramIdOrNameQuery>,
+    with_code: web::Query<WithCodeQuery>,
 ) -> Result<HttpResponse, ManagerError> {
-    let program_id = ProgramId(parse_uuid_param(&req, "program_id")?);
+    let with_code = with_code.with_code.unwrap_or(false);
+    if let Some(id) = req.id {
+        let program = state
+            .db
+            .lock()
+            .await
+            .get_program_by_id(*tenant_id, ProgramId(id), with_code)
+            .await?;
 
-    Ok(state
-        .db
-        .lock()
-        .await
-        .program_code(*tenant_id, program_id)
-        .await
-        .map(|(program, code)| {
-            HttpResponse::Ok()
-                .insert_header(CacheControl(vec![CacheDirective::NoCache]))
-                .json(&ProgramCodeResponse { program, code })
-        })?)
+        Ok(HttpResponse::Ok()
+            .insert_header(CacheControl(vec![CacheDirective::NoCache]))
+            .json(&vec![program]))
+    } else if let Some(name) = req.name.clone() {
+        let program = state
+            .db
+            .lock()
+            .await
+            .get_program_by_name(*tenant_id, &name, with_code)
+            .await?;
+        Ok(HttpResponse::Ok()
+            .insert_header(CacheControl(vec![CacheDirective::NoCache]))
+            .json(&vec![program]))
+    } else {
+        let programs = state
+            .db
+            .lock()
+            .await
+            .list_programs(*tenant_id, with_code)
+            .await?;
+        Ok(HttpResponse::Ok()
+            .insert_header(CacheControl(vec![CacheDirective::NoCache]))
+            .json(&programs))
+    }
 }
 
-/// Returns program descriptor, including current program version and
-/// compilation status.
+/// Fetch a program by ID.
```

Comment:
This does seem redundant since the previous endpoint can be used for the same purpose.

Reply:
One returns a list (with potentially one item when the search criterion of ?id or ?name is used), whereas the other returns the program. I find having both a bit more convenient (doesn't require clients to extract a list, check if it has one entry etc).

Reply:
Maybe I should just remove the ?id search option, which is indeed redundant.

Reply:
Not sure I follow, but I prefer being able to retrieve a single program, not in a list form. Code is more readable if it can check for null/404 rather than having to check for a empty list.

Reply:
@gz there's `GET /programs/{program_id}` which will retrieve a single program. There's also `GET /programs`, which returns a list, and optionally can be narrowed down with search terms `?id` and `?name`.